lib/pulse_reader/pulse_reader.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <stm32wbxx_ll_dma.h>
+#include <stm32wbxx_ll_dmamux.h>
+#include <stm32wbxx_ll_tim.h>
+#include <stm32wbxx_ll_exti.h>
+
+#include <furi_hal_gpio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define PULSE_READER_NO_EDGE 0xFFFFFFFFUL
+#define PULSE_READER_LOST_EDGE 0xFFFFFFFEUL
+#define F_TIM2 64000000UL
+
+/**
+ * unit of the edge durations to return
+ */
+typedef enum {
+    PulseReaderUnit64MHz,
+    PulseReaderUnitPicosecond,
+    PulseReaderUnitNanosecond,
+    PulseReaderUnitMicrosecond,
+} PulseReaderUnit;
+
+typedef struct {
```

Comment:
Do you need to expose this structure here? Can it be anonymous?

Reply:
it is required here:
![grafik](https://user-images.githubusercontent.com/479231/209866338-3e2d006e-2f4c-4e51-a4da-0c63b0a402c8.png)

Reply:
> digital_signal_update_dma
In fact the asm version is faster than any C version.
It prepares all data to be written into registers, which get written into memory using a single-fetch, multiple write STM operation right after the "transaction finished" bit flipped.

Speaking in pipelining, there is just a single FETCH and DECODE pipeline step, but only multiple cycles in EXECUTE step.
This was required to get even NfcA emulation working fully in software.
Without this optimization, the short pulses in NfcA modulation were too short to set up the next DMA transaction.

Reply:
It's not only about performance but also about portability. Please provide C version alongside so we'll be able to compare them ;-)

Reply:
Sure, will provide it.
But keep in mind that the key is the latency between "DMA finished" and "next DMA started" is the key.
![grafik](https://user-images.githubusercontent.com/479231/209870486-fd898962-0332-4811-8785-0697a3a247ef.png)

If the calculations I did are correct, we have 37 clock cycles from a finished DMA to DMA being started again.
Minus some triggering delays. This caused emulating NfcA this way not to work if I used some low level C code.